components/ord-service/src/main/java/com/sap/cloud/cmp/ord/service/token/Token.java

```diff
@@ -40,23 +41,42 @@ public class Token {
 
     private Set<String> formationIDsClaims;
 
-    public Token(SubscriptionHelper subscriptionHelper, String idTokenEncoded) throws JsonMappingException, JsonProcessingException {
+    public String applicationTenantId;
+
+    public Token(SubscriptionHelper subscriptionHelper, String idTokenEncoded, String applicationTenantId) throws JsonMappingException, JsonProcessingException {
         this.subscriptionHelper = subscriptionHelper;
         String idTokenDecoded = decodeIDToken(idTokenEncoded);
         this.content = mapper.readTree(idTokenDecoded);
         this.formationIDsClaims = new HashSet<>();
+        this.applicationTenantId = applicationTenantId;
     }
 
     public String extractTenant() {
         String unescapedTenants = content.get(TENANTS_MAP_KEY).asText().replace("\\", "");
 
         String providerTenantID = "";
         try {
+            SelfRegisteredRepository repo = subscriptionHelper.getRepo();
+
             JsonNode tenantsTree = mapper.readTree(unescapedTenants);
 
             String tenant = tenantsTree.path(CONSUMER_TENANT_KEY).asText();
             providerTenantID = tenantsTree.path(PROVIDER_TENANT_KEY).asText();
             if (providerTenantID == null || providerTenantID.isEmpty() || providerTenantID.equals(tenant)) {
+                if (applicationTenantId != null && !applicationTenantId.isEmpty()) {
+                    logger.info("Application local tenant ID is provided through header");
+                    String consumerID = content.get(CONSUMER_ID_KEY).asText();
+                    if (consumerID == null || consumerID.isEmpty()) {
+                        logger.error("consumer ID could not be empty");
+                        return tenant;
```

Comment:
Do you think it will be better to return explicitly `""` here because in this if `tenant` will always be empty and it will be more intuitive in the future while reading the code?

Reply:
The code was adapted to cover some corner cases